pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -600,17 +600,30 @@ class Timestamp(_Timestamp):
             value = self.tz_localize(None).value
         else:
             value = self.value
+
+        # GH 19206
+        # to deal with round-off when unit is large
+        if unit >= 1e9:
+            divisor = 10 ** int(np.log10(unit / 1e7))
+        else:
+            divisor = 1
+
         if unit < 1000 and unit % 1000 != 0:
             # for nano rounding, work with the last 6 digits separately
             # due to float precision
```

Comment:
can this case be incorporated as well in the same divisor check?

Reply:
maybe set both divisor & buff in a single if that handles really small, really large and other units (divisor=1, buff=1)

Reply:
can you do this?

Reply:
I don't see a simple way of doing this. We would essentially have to move the logic until the arithmetic, which I think would make the code more opaque.  I can move the divisor check into the case were unit >= 1000 and refactor some of it to make it a bit cleaner.

Reply:
that's ok, just trying to make it more clear.